File: c_files/raylib/bouncing_balls_v2.c
#include <raylib.h>
#include <raymath.h>
#include <math.h>

#define NUM_OF_BALLS 200

typedef struct Ball
{
  Vector2 pos;
  float radius;
  Vector2 velocity;
  Color color;

} Ball;

int main(void)
{
  const int SCREEN_WIDTH = 800;
  const int SCREEN_HEIGHT = 600;
  const char SCREEN_TITLE[] = "Bouncing Balls";
  const Color SCREEN_BACKGROUND = SKYBLUE;

  InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

  Ball balls[NUM_OF_BALLS];

  for (int i = 0; i < NUM_OF_BALLS; i++)
  {
    float speed = GetRandomValue(100, 300);
    float angle = GetRandomValue(0, 360) * DEG2RAD;

    Ball ball = {
        .radius = GetRandomValue(5, 30),
        .pos.x = GetRandomValue(ball.radius, GetScreenWidth() - ball.radius),
        .pos.y = GetRandomValue(ball.radius, GetScreenHeight() - ball.radius),
        .color =
            (Color){
                GetRandomValue(0, 255),
                GetRandomValue(0, 255),
                GetRandomValue(0, 255),
                255,
            },
        .velocity.x = cosf(angle) * speed,
        .velocity.y = sinf(angle) * speed,
    };

    balls[i] = ball;
  }

  while (!WindowShouldClose())
  {
    for (int i = 0; i < NUM_OF_BALLS; i++)
    {
      // move ball
      balls[i].pos.x += balls[i].velocity.x * GetFrameTime();
      balls[i].pos.y += balls[i].velocity.y * GetFrameTime();

      // ball bounds
      if (balls[i].pos.x < balls[i].radius ||
          balls[i].pos.x > GetScreenWidth() - balls[i].radius)
      {
        balls[i].velocity.x *= -1;
      }
      
      if (balls[i].pos.y < balls[i].radius ||
          balls[i].pos.y > GetScreenHeight() - balls[i].radius)
      {
        balls[i].velocity.y *= -1;
      }
    }

    BeginDrawing();
    ClearBackground(SCREEN_BACKGROUND);

    for (int i = 0; i < NUM_OF_BALLS; i++)
    {
      DrawCircleV(balls[i].pos, balls[i].radius, balls[i].color);
    }

    EndDrawing();
  }

  CloseWindow();

  return 0;
}

File: python_files/raylib_redo/bouncing_balls_v2.py
import pyray as p
from math import cos, sin, radians

SCREEN_WIDTH: int = 800
SCREEN_HEIGHT: int = 600
SCREEN_TITLE: str = "Bouncing Balls"
SCREEN_BACKGROUND: p.Color = p.SKYBLUE

NUM_OF_BALLS: int = 200


class Ball:
    def __init__(self) -> None:
        self.radius: float = p.get_random_value(10, 30)
        self.pos: p.Vector2 = p.Vector2(
            p.get_random_value(self.radius, p.get_screen_width() - self.radius),
            p.get_random_value(self.radius, p.get_screen_height() - self.radius),
        )

        speed: float = p.get_random_value(100, 300)
        angle: float = p.get_random_value(1, 360)
        self.velocity: p.Vector2 = p.Vector2(
            speed * cos(radians(angle)),
            speed * sin(radians(angle)),
        )
        
        self.color: p.Color = p.color_alpha(
            p.Color(
                p.get_random_value(0, 255),
                p.get_random_value(0, 255),
                p.get_random_value(0, 255),
            ),
            255,
        )

    def draw(self) -> None:
        p.draw_circle_v(self.pos, self.radius, self.color)

    def update(self) -> None:
        # move ball
        self.pos.x += self.velocity.x * p.get_frame_time()
        self.pos.y += self.velocity.y * p.get_frame_time()

        # ball bounds
        if self.pos.x < self.radius or self.pos.x > p.get_screen_width() - self.radius:
            self.velocity.x *= -1
        if self.pos.y < self.radius or self.pos.y > p.get_screen_height() - self.radius:
            self.velocity.y *= -1


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    balls: list[Ball] = []

    for i in range(NUM_OF_BALLS):
        ball: Ball = Ball()

        balls.append(ball)

    while not p.window_should_close():
        for ball in balls:
            ball.update()

        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        for ball in balls:
            ball.draw()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()
